### ISSUE TITLE
worker logs should be regularly written as a task is running

### ISSUE THREAD

**Issue by bhearsum** [taskcluster/taskcluster#6685]
When tasks resolve with an exception, it is very common that they have no logs available. Worse still, there is typically a `live.log` artifact that always shows a `NetworkError` (presumably because it is still looking to grab the log through the proxy).

One such example is https://firefox-ci-tc.services.mozilla.com/tasks/DZvVQ-VUTPSyPBBS13Bwfg/runs/1, which run for 11 hours and most certainly generated a lot of useful log output before getting preempted.

It would be swell if whatever partial log exists would be made available when a task is resolved with exception.

There was some chatter on Matrix about maybe updating an S3 artifact to make this possible, but I also naively wonder if we could be flushing it to disk regularly in the artifact directory, such that it will be uploaded along with other artifacts in many cases? (Perhaps in addition to S3 posts?)

**Comment by petemoore** [comment]
It occurs to me, that in the short term, this is addressable at a task level (or even via taskgraph). Tasks can publish artifacts dynamically, while the task is running. So there is nothing to stop a task from writing its output to both the console, and to a backing file, which it routinely pulbishes as an artifact, e.g. every time it has grown to 1MB in size, etc.
This might not be an ideal permanent solution, but it does highlight that taskcluster provides a great deal of flexibility here for tasks to publish whatever they like, on whatever cadence they see fit. In the absence of having an artifact type that doesn't exist today, this could be achieved with the existing s3 artifact type and some calls to queue.createArtifact while the task is running, using the taskcluster proxy.

**Comment by bhearsum** [comment]
I agree that this helps with some cases, but unless I'm mistaken it does _not_ help in cases where the worker disappears without being able to upload artifacts, which is the main case I filed this about.

**Comment by marco-c** [comment]
@bhearsum I might be missing something, but if the worker periodically uploads logs to S3, e.g. once every 10 minutes, then a disappearing worker should not be a problem. E.g. if the worker runs for 11 hours like in the example above, we will have uplaoded the artifacts at 10 hours and 50 minutes.

**Comment by bhearsum** [comment]
> @bhearsum I might be missing something, but if the worker periodically uploads logs to S3, e.g. once every 10 minutes, then a disappearing worker should not be a problem. E.g. if the worker runs for 11 hours like in the example above, we will have uplaoded the artifacts at 10 hours and 50 minutes.

Yes, I agree. My most recent comment was in reply to @petemoore's comment. Unless I misunderstood something in it (always possible...), the idea wouldn't ensure files are published unless the task completes successfully, or a termination is handled gracefully.

**Comment by djmitche** [comment]
I think @petemoore is suggesting uploading an artifact at some interval _during_ execution of the task, by calling `queue.createArtifact` etc. Once that's complete, the artifact exists, regardless of what subsequently happens to the worker.

The downside is that an artifact cannot be modified once it is written, so for example if logs were uploaded every MB you might have `partial-1MB.log`, `partial-2MB.log`, `partial-3MB.log`, etc. You're also paying for upload bandwidth for those files, as well as storage (although the storage might only be for a few days).

Would it be useful if a worker could write logs to a log-handling service (logstash, elk, splunk, whatever) instead of an artifact? That would accomplish the streaming pretty nicely, although it would take it out of Taskcluster's orbit. Maybe the worker could write a Redirect artifact with a URL pointing at the log service's UI -- perhaps even creating that artifact when the task starts.
